hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/SparkBroadcastManager.java

```diff
@@ -0,0 +1,121 @@
+  protected Broadcast<SQLConf> sqlConfBroadcast;
+  protected Broadcast<SparkParquetReader> parquetReaderBroadcast;
+  protected Broadcast<SerializableConfiguration> configurationBroadcast;
+
+  public SparkBroadcastManager(HoodieEngineContext context) {
+    this.context = context;
+  }
+
+  // Prepare broadcast variables.
+  @Override
+  public void prepareAndBroadcast() {
+    // This needs to be fixed.
+    if (!(context instanceof HoodieSparkEngineContext)) {
+      throw new HoodieIOException("Expected to be called using Engine's context and not local context");
+    }
+
+    HoodieSparkEngineContext hoodieSparkEngineContext = (HoodieSparkEngineContext) context;
+    SQLConf sqlConf = hoodieSparkEngineContext.getSqlContext().sessionState().conf();
+    JavaSparkContext jsc = hoodieSparkEngineContext.jsc();
+
+    // TODO: Confirm what is the correct way to set this config.
+    boolean returningBatch = sqlConf.parquetVectorizedReaderEnabled();
+    scala.collection.immutable.Map<String, String> options =
+        scala.collection.immutable.Map$.MODULE$.<String, String>empty()
+            .$plus(new Tuple2<>(FileFormat.OPTION_RETURNING_BATCH(), Boolean.toString(returningBatch)));
+
+    // Do broadcast.
+    sqlConfBroadcast = jsc.broadcast(sqlConf);
+    configurationBroadcast = jsc.broadcast(new SerializableConfiguration(jsc.hadoopConfiguration()));
+    // TODO: Disable vectorization as of now. Assign it based on relevant settings.
+    // TODO: Verify if we can construct the reader on the executor side if we has broadcast all necessary variables.
+    parquetReaderOpt = Option.of(SparkAdapterSupport$.MODULE$.sparkAdapter().createParquetFileReader(
+        false, sqlConfBroadcast.getValue(), options, configurationBroadcast.getValue().value()));
+    parquetReaderBroadcast = jsc.broadcast(parquetReaderOpt.get());
+  }
+
+  @Override
+  public Option<HoodieReaderContext> retrieveFileGroupReaderContext(StoragePath basePath) {
+    if (parquetReaderBroadcast == null) {
+      LOG.warn("ParquetReader is not broadcast; cannot create file group reader");
+      return Option.empty();
+    }
+
+    // Reconstruct metaClient, since underlying properties may not be passed from driver to executors.
+    HoodieTableMetaClient metaClient = HoodieTableMetaClient.builder()
+        .setBasePath(basePath)
+        .setConf(new HadoopStorageConfiguration(configurationBroadcast.getValue().value())).build();
+    SparkParquetReader sparkParquetReader = parquetReaderBroadcast.getValue();
+    if (sparkParquetReader != null) {
+      List<Filter> filters = new ArrayList<>();
+      return Option.of(new SparkFileFormatInternalRowReaderContext(
+          sparkParquetReader,
+          JavaConverters.asScalaBufferConverter(filters).asScala().toSeq(),
+          JavaConverters.asScalaBufferConverter(filters).asScala().toSeq()));
+    } else {
+      LOG.warn("ParquetFileReader is null");
```

Comment:
also, should we throw exception in this case?

Reply:
Yes.  Fixed.